### ISSUE TITLE
.forward "hangs" in a flask app.

### ISSUE THREAD

**Issue by GSanchis** [awslabs/keras-apache-mxnet#149]
Dear all,

I must admit I'm not too sure about how to help dig deeper into the issue I am facing. Here goes a quick description:
- Running python3.5, keras-mxnet 2.2.0, keras 2.2.2, mxnet 1.0.0 (although I have tried also with mxnet 1.2.1)
- I'm building a flask app for serving the model, written in Keras, which includes a convolutional network for classifying text. Inputs are chars/words/BPE tokens, in one-shot representation, output is a dense layer.
- I have a model defined and loaded outside the predict endpoint, so that subsequent calls to the predict endpoint do not require to reload the model each time.
- When querying the predict endpoint, the first time around everything works fine. The second time, it gets stuck in the predict code and never comes back. Delving deep into the code of keras-mxnet, I found out that it gets stuck in the predict function defined in `_make_predict_function`, in line
```                self._module.forward(batch, is_train=False)```
(line 4840 of the current code, so basically, in the `forward` step). 
Context: https://github.com/awslabs/keras-apache-mxnet/blob/332c62f7b521b109431c0ef91317b2ce00e4cfb7/keras/backend/mxnet_backend.py#L4840

Now. The interesting part is that this problem arises when using the latest Flask (1.0.2). However, using Flask 0.12.4 does not lead to this problem, and the predict endpoint can be queried as often as required.

We had a similar problem in the past using Keras with a TensorFlow backend, although that seemed to be due to session problems, and that's why we switched to MXNet. 

I'm not providing a simple code to reproduce the problem because, this being included in a Flask app, I'm not too sure what a "minimal code" would be, aside from the full code base.

Any clues regarding what might be going on?

Thank you very much in advance!

**Comment by roywei** [comment]
@GSanchis Thanks for submitting his issue. We will take a look.
Meanwhile, for model serving you can try to save the model in mxnet format, and use other methods or language bindings that are more suitable for serving.
[tutorial to save mxnet model](https://github.com/awslabs/keras-apache-mxnet/blob/master/docs/mxnet_backend/save_mxnet_model.md)
[ mxnet model server]( https://github.com/awslabs/mxnet-model-server)
[keras-mxnet + model server](https://medium.com/apache-mxnet/deploy-a-smile-detector-with-keras-mxnet-and-mxnet-model-server-48cd9741b6d2)
[keras-mxnet + scala](https://medium.com/apache-mxnet/train-using-keras-mxnet-and-inference-using-mxnet-scala-api-49476a16a46a)

**Comment by GSanchis** [comment]
Thanks for that, @roywei . In fact, I had alrady used the MMS in another project about building a recommender system on top of MXNet, and it works beautifully. However, given that in this project we started off with a TensorFlow backend and then switched to the MXNet backend I had not even thought about it, perhaps also because I had not understood that I could actually save the model to MXNet format.

One question I would have, however, about the post that explains how to save the MXNet model after training: one of the things I like about Keras is that you can enable a callback to save the weights that perform the best in the validation set. Correct me if I'm wrong, but using `save_mxnet_model()` would actually save the weights after the full training (and not the best validation epoch weights), right? This might only have a minor impact, or not: keeping the best-performing weights in validation seems to me the right thing to do.

In any case, let me know if you need any other info or if I can help in any way to solve this issue: having worked with MXNet for quite some time now, I was very surprised to fall into this problem.

**Comment by sandeep-krishnamurthy** [comment]
@GSanchis - You are spot on! Recently, `MXNetModelCheckpoint` is added that helps in checkpointing models (MXNet model) similar to Keras ModelCheckpoint.
It is currently `dev2` branch. https://github.com/awslabs/keras-apache-mxnet/blob/dev2/docs/mxnet_backend/save_mxnet_model.md#checkpoint-mxnet-model-using-mxnetmodelcheckpoint-callback

We are preparing for 2.2.2 release that will include MXNetModelCheckpoint

**Comment by GSanchis** [comment]
That's great news, @sandeep-krishnamurthy !! Thanks for the tip. I'll keep an eye on that. 

Since we're at it - and this is fully unrelated to the original thread so feel free to ignore this - there's another reason for not using MMS in this case, just in case there is also a good solution: the API needs to feature training and prediction endpoints. Given that, as far as I know, MMS only serves trained models, I would still require to set up a Flask server for the training endpoint. This implies that either I serve both endpoints through different (potentially non-standard) ports, or I need an Nginx (or the similar) to route the requests to the appropriate server. Given that we typically work with Docker containers, in the end the setup with MMS turns out much more complex than the setup with a simple Flask. Am I missing something?

**Comment by sandeep-krishnamurthy** [comment]
@GSanchis - Interesting usecase and I am not very clear on why do you need an endpoint for training? Are you thinking about Online learning type usecase where your model is continuously being trained?

**Comment by sandeep-krishnamurthy** [comment]
You should definitely consider opening an issue here - https://github.com/awslabs/mxnet-model-server/issues to talk about your usecase and potentially model server contributors can suggest solution or take up a feature request

**Comment by sandeep-krishnamurthy** [comment]
@GSanchis Thanks for the details.
Model Server is used for online inference (Synchronous one request or mini-batch request). Hence, I wanted to understand why and how would training end point in the model server would work synchronously and for what use cases. 

Please correct me if I am wrong, the usecases you explained is more like an asynchronous call that initiates a batch offline job.

Yes, please, this feature request discussion in MMS will be very helpful. Thanks.

Coming back to keras-mxnet :-) Did you get a chance to pull `dev2` branch and try MXNetModelCheckpoint?

**Comment by kalyc** [comment]
@GSanchis thanks for opening this issue. We hope that you have had a chance for trying out MXNetModelCheckpoint. Its available in Keras-MXNet v2.2.2 release now. Closing this issue at the moment. Please feel free to re-open if there are any other questions at your end.
